packages/core/src/util/renderer.ts

```diff
@@ -310,37 +330,49 @@ const conditionallyRequired = (
   }, nestedAllOfSchema);
 };
 
+const getNextHigherSchemaPath = (schemaPath: string): string => {
+  const pathSegments = schemaPath.split('/');
+  const lastSegment = pathSegments[pathSegments.length - 1];
+
+  // We'd normally jump two segments back, but if we're in an `items` key, we want to check its parent
+  // Example path: '#/properties/anotherObject/properties/myArray/items/properties/propertyName'
+  const nextHigherSegmentIndexDifference = lastSegment === 'items' ? 1 : 2;
+  const nextHigherSchemaSegments = pathSegments.slice(
+    0,
+    pathSegments.length - nextHigherSegmentIndexDifference
+  );
+
+  return nextHigherSchemaSegments.join('/');
+};
+
+const getNextHigherDataPath = (dataPath: string): string => {
+  const dataPathSegments = dataPath.split('.');
+  return dataPathSegments.slice(0, dataPathSegments.length - 1).join('.');
+};
+
 /**
  * Check if property is being required in the parent schema
  */
 const isRequiredInParent = (
   schema: JsonSchema,
-  rootSchema: JsonSchema,
-  path: string,
+  schemaPath: string,
   segment: string,
   prevSegments: string[],
-  data: Record<string, unknown>
+  data: Record<string, unknown>,
+  dataPath: string
 ): boolean => {
-  const pathSegments = path.split('/');
-  const lastSegment = pathSegments[pathSegments.length - 3];
-  const nextHigherSchemaSegments = pathSegments.slice(
-    0,
-    pathSegments.length - 4
-  );
+  const pathSegments = schemaPath.split('/');
+  const lastSegment = pathSegments[pathSegments.length - 1];
+  const nextHigherSchemaPath = getNextHigherSchemaPath(schemaPath);
 
-  if (!nextHigherSchemaSegments.length) {
+  if (!nextHigherSchemaPath) {
     return false;
   }
 
-  const nextHigherSchemaPath = nextHigherSchemaSegments.join('/');
+  const nextHigherSchema = Resolve.schema(schema, nextHigherSchemaPath, schema);
```

Comment:
I don't think using `schema` instead of `rootSchema` is better here. From what I see, it is true that we don't need the `rootSchema`, but if a new feature is added that relies on it, maybe it would cause a problem.